\documentclass[a4paper, 10pt]{article}        %\documentclass{} slides report proc book 
\usepackage[margin=1in]{geometry}
\usepackage{indentfirst}
\usepackage[]{graphicx}
\usepackage[]{longtable}
\usepackage[hyperref]{xcolor}
\definecolor{teal}{RGB}{0, 128, 128}
\usepackage{hyperref, xcolor}
\usepackage{standalone}
\usepackage{pdflscape}


\hypersetup{
        colorlinks = true,
        allcolors={teal},
        linkcolor={teal}
       %urlcolor=[rgb]{0,128,128},
     }
\title{Spec for query region -by alias}     %query by annotation

\begin{document}
\date{}
\maketitle
%\author{{\color{teal}Claudia Vasallo}}

\section{Definition}

%This query uses aliases to refer to genomic regions that will be mapped to a \texttt{query by region -(by coordinates)}  that is already implemented. The mapping of these aliases to \texttt{query by region} parameters (\textbf{refseqId} and \textbf{start}, \textbf{end} positions) comes from genome annotation file. 

This query will use aliases to refer to genomic regions coordinates that will be mapped to a \texttt{query by region -(by coordinates)}, so it can be considered a special case of \texttt{query by region} and be implemented on top of it. The mapping of names/aliases to \texttt{query by region} parameters (\textbf{refseqId} and \textbf{start}, \textbf{end} positions) comes from genome annotation file. 

%For \texttt{query by region -(by name/alias)} in Viral Beacon NCBI RefSeq annotation for SARS-CoV2 reference sequence (NC\_045512.2) will be used in backend as a dictionary. This annotation file contains the RefSeqID ("NC\_045512.2") and the region coordinates that can be given to dev-team as a table like in table \ref{table:genomic}- \ref{table:matpt}. In frontend either a dropdown menu or a suggesting-while-typing menu should appear based on the dictionary region names like in figure \ref{fig:querybox}.

Note: This query type implementation could be deprecated after we have better annotated VCFs that include mature proteins and possibly other functional domains. Then, \texttt{query by annotation}, an implementation that uses variant annotations in VCF regardless of RefSeqID, can substitute for this.


 %QUERY
\section{Query specifications}

\noindent{Parameters: same as in \texttt{query by region}, i.e \textbf{assemblyId} and/or \textbf{refseqId}, \textbf{start} and \textbf{end}.}
%Parameters for this query are \textbf{assemblyId} and/or \textbf{refseqId} and an alias substituting for \textbf{start} and \textbf{end} in \texttt{query by region}.

For the time being \textbf{refseqId} is set to default value since we only have one virus (taxonId:2697049 ) with just one RefSeq (refseqId:NC\_045512.2) in Viral Beacon. Note: When/if more viruses come (and also for generic Beacon) user will have to enter those parameters as well (so there should be a table in backend mapping all available \textbf{taxonId} and their available \textbf{assemblyId} and other table mapping assemblyIds to their corresponding sets of \textbf{refseqId}).\\

%So, either default or selected, \textbf{assemblyId} or \textbf{refseqId} are set.% previous to query.

\noindent{User input: \textbf{assemblyId} or \textbf{refseqId} (default now) and an alias substituting for \textbf{start} and \textbf{end} in \texttt{query by region}.}\\

\noindent{Frontend suggestion: User enters a genomic region name/alias/identifier/accession in query box like in figure \ref{fig:querybox}. Either a dropdown menu or a suggesting-while-typing menu should appear showing the query 'options' based on the mapping table stored in backend for this \textbf{refseqId}.}
 
     \begin{figure}[h!]
     \centering
    %\includegraphics[width=0.75\textwidth]{query_by_region_box.pdf}
     \includegraphics[width=0.75\textwidth]{query_by_region_box_3.pdf}
     \caption{Query by region box}
     \label{fig:querybox}
     \end{figure}
     
(Later, when we have implementation for multiple \texttt{query by region} at once -as for \texttt{query by motif}- the idea would be being able to allow search also for lists of those aliases (e.g entered separated by commas), and for grouping terms such as "intergenic", "coding", "non-coding", "utr", "stem loops" or "structural protein", that would translate into a list of start:end coordinates)


% PREREQUISITES
\section{Prerequisites} 
\begin{itemize}
%\item[-] A table/dictionary mapping names/aliases/accessions accepted in this field to genomic region coordinates (a plain text file looking like tables below, containing reference genome annotation- coordinates for genes, UTRs, stem loops, CDSs and mature proteins), given by bio-team.

\item[1] A table/dictionary mapping names/aliases or synonyms/accessions accepted in this field to genomic region coordinates. \\
To do this, the genomic annotation file for the specified \textbf{refseqId} was fetched. For SARS-CoV2 we will use reference sequence NCBI RefSeq NC\_045512.2 annotation available at \href{https://www.ncbi.nlm.nih.gov/nuccore/NC\_045512.2?report=gbwithparts&log$=seqview}{NCBI web}.
%(relevant rows are here \href{https://github.com/clauw87/virusbeacon/blob/raw_ideas/annot_file.txt}{annot.file\_txt}).
This annotation contains the \textbf{refseqId} ("NC\_045512.2") and the names/ aliases of genomic regions with their corresponding \textbf{start} and \textbf{end} coordinates. \\
The relevant data was parsed as a table like in table \ref{table:1} (download the table as \href{https://github.com/clauw87/virusbeacon/blob/raw_ideas/annot_coord_table.txt}{csv}).
\end{itemize}

% galaxy doesn't have stem loops 
%\href{https://workshop.usegalaxy.org/datasets/1d19a67d0fa8e3d4/display/?preview=True}{galaxy}


%other name/aliases are in column alias and ids i column id should be findable as well and showed as name.
%Annotations for 
%coding regions > 11 genes, 12 CDSs (protein products), 26 mature peptides (processed proteins)
%non coding regions > 2 utrs
%functional > structural: 5 stem loops


%	TO DO
%links to files >
%original gff
%my refseq> same info on gff plus mature peptides.
%GALAXY down


% IMPLEMENTATION
\section{Implementation} 
\begin{itemize}
\item[1] Entries from table should be converted to 'options' for frontend. The 'option' would work as a dictionary key for all names/aliases in the table row. The 'option' is what should appear as suggestions while user types a part of either \texttt{name}, \texttt{syn\_alias},  \texttt{locus\_tag}, \texttt{id} or \texttt{accession} of its row, so some flexibility is allowed. Alternatively, 'options' will appear in a dropdown menu (so, no flexibility but user knows his options).

'Options' will be constructed by concatenating 'value in \underline{\texttt{class}}:value in \underline{\texttt{name}}' from table \ref{table:1}, eg. gene: ORF8, cds: ORF1a polyprotein, functional: RNA-dependent RNA polymerase, functional: Coronavirus 3' UTR pseudoknot stem-loop 1, non-coding: 5'UTR.



%The 'option' is the name that should appear as suggestions in search bar while user is typing either \texttt{name}, \texttt{syn\_alias}, \texttt{id} or \texttt{accession}.



\item[2] Coordinates (\texttt{start} and \texttt{end}) will be pulled from table \ref{table:1} using the 'option' \texttt{name} and then \texttt{query by region -(by coordinates)} is run as usual using these as parameters \texttt{start} and \texttt{end}.

\end{itemize}




\input{annot_coord_table}
%\caption{Table for coordinates- name/aliases}
%\label{table:1}




\section{Response}
Response will be same as that of a \texttt{query by region}.
%\subsection{How to show many query by regions}
%\subsection{Statistics in genomic region}
%\begin{itemize}
%\item[-] Freq variants per variantType
%\item[-] dN/dS in region
%\end{itemize}
%\subsection{Statistics per position within region}
%\begin{itemize}
%\item[-] Needle for region freq alternate per position
%\item[-] Freq variants per variantType
%\item[-] dN/dS per position
%\end{itemize}

%\subsection{What is necessary} 

\end{document}




